src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -346,11 +343,35 @@ private bool IsLegacyAdalCacheEnabled(AuthenticationRequestParameters requestPar
             }
         }
 
-        private DateTimeOffset CalculateSuggestedCacheExpiry()
+        /// <summary>
+        /// Important note: we should not be suggesting expiration dates that are in the past, as it breaks some cache implementations.
+        /// </summary>
+        internal /* for testing */ static DateTimeOffset? CalculateSuggestedCacheExpiry(
+            ITokenCacheAccessor accessor, 
+            ICoreLogger logger)
         {
-            IEnumerable<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensWithNoLocks(true);
-            DateTimeOffset cacheExpiry = tokenCacheItems.Max(item => item.ExpiresOn);
-            return cacheExpiry;
+            // If we have refresh tokens in the cache, we cannot suggest expiration
+            // because refresh token expiration is not disclosed to SDKs and RTs are long lived anyway (3 months by default)
+            if (accessor.GetAllRefreshTokens().Count == 0)
+            {
+                IReadOnlyList<MsalAccessTokenCacheItem> tokenCacheItems = accessor.GetAllAccessTokens(optionalPartitionKey: null);
+                if (tokenCacheItems.Count == 0)
+                {
+                    logger.Warning("[CalculateSuggestedCacheExpiry] No access tokens or refresh tokens found in the accessor. Not returning any expiration.");
+                    return null;
+                }
+
+                DateTimeOffset cacheExpiry = tokenCacheItems.Max(item => item.ExpiresOn);
+                
+                // do not suggest an expiration date from the past or within 5 min, as tokens will not be usable anyway
+                // and HasTokens will be set to false, letting implementers know to delete the cache node
+                if (cacheExpiry < DateTimeOffset.UtcNow + Constants.AccessTokenExpirationBuffer)
```

Comment:
This fix looks good for us. I know one 1P who hit this issue, and they could test the ID web version for us, if you want. Also, I think you are spending too much time in SAL, as I see you remove the `{ }` on the one liners ;)